README.md
# FiiSoft Molecular Weight Calculator

Simple molecular weight calculator. Allows you to calculate masses of chemical particles based on stoichiometric formulas.

See tests to know what kind of formulas this package can handle.

My advice is - do not use it unless you are enough strong mentally to immune for such bad code.

CHANGELOG.md
# Changelog

All important changes to `fiisoft-molecular-weight-calc` will be documented in this file

## 1.1.0

Added second source of data: atomic masses computed from data in pdf files published by IUPAC on page www.isotopesmatter.com.

## 1.0.0

First version of package.

tests/MolecularWeightCalculatorTest.php
<?php

namespace FiiSoft\Test\MolecularWeightCalculator;

use FiiSoft\MolecularWeightCalculator\MolecularWeightCalculator;

class MolecularWeightCalculatorTest extends \PHPUnit_Framework_TestCase
{
    /** @var MolecularWeightCalculator */
    private $calc;
    
    protected function setUp()
    {
        parent::setUp();
        
        $this->calc = new MolecularWeightCalculator();
    }
    
    public function test_get_data()
    {
        self::assertInternalType('array', $this->calc->getData());
    }
    
    public function test_it_can_compute_molecural_weights_from_various_chemical_formulas()
    {
        foreach ($this->getTestData() as $chemicalFormula => $expectedWeight) {
            $weight = $this->calc->computeWeight($chemicalFormula);
            self::assertTrue(
                abs($expectedWeight - $weight) < 0.0001,
                'test: '.$chemicalFormula.', expected weight: '.$expectedWeight.', computed weight: '.$weight
            );
        }
    }
    
    /**
     * @return array
     */
    private function getTestData()
    {
        return [
            'CaFeBiBa' => 40.078022556685 + 55.84514413384 + 208.9804 + 137.3266716329,
            'Na2CO3' => 2 * 22.98976928 + 12.010735896734 + 3 * 15.999404924748,
            'C6H12O6' => 6 * 12.010735896734 + 12 * 1.0079407540258 + 6 * 15.999404924748,
            'NH4Cl' => 14.006703211018 + 4 * 1.0079407540258 + 35.453,
            'Ca(OH)2' => 40.078022556685 + 2 * (15.999404924748 + 1.0079407540258),
            'Mg2(C2H4(OH)2)3SO4' => 2 * 24.30505162813 + 3 * (2 * 12.010735896734 + 4 * 1.0079407540258 + 2 * (15.999404924748 + 1.0079407540258)) + 32.064787405608 + 4 * 15.999404924748,
            'C11H19N3O6S' => 11 * 12.010735896734 + 19 * 1.0079407540258 + 3 * 14.006703211018 + 6 * 15.999404924748 + 32.064787405608,
            'H15-N3-C12-O4-S2-Cl' => 15 * 1.0079407540258 + 3 * 14.006703211018 + 12 * 12.010735896734 + 4 * 15.999404924748 + 2 * 32.064787405608 + 35.453,
            'K[AlSi3O8]' => 39.0983 + 26.9815386 + 3 * 28.0855 + 8 * 15.999404924748,
            'Ca2(Fe2+)4Al[(Si7Al)O22(OH)2]' => 2 * 40.078022556685 + 4 * 55.84514413384 + 26.9815386 + 7 * 28.0855 + 26.9815386 + 22 * 15.999404924748 + 2 * (15.999404924748 + 1.0079407540258),
            'CaFe7Mg2B3Si6O31H3' => 40.078022556685 + 7 * 55.84514413384 + 2 * 24.30505162813 + 3 * 10.811 + 31 * 15.999404924748 + 6 * 28.0855 + 3 * 1.0079407540258,
            'Ca(Fe)2Fe(Fe)4Mg2(BO3)3Si6O18(OH)3O' => 40.078022556685 + 7 * 55.84514413384 + 2 * 24.30505162813 + 3 * 10.811 + 31 * 15.999404924748 + 6 * 28.0855 + 3 * 1.0079407540258,
            'Ca[(Fe)2Fe][(Fe)4Mg2](BO3)3Si6O18(OH)3O' => 40.078022556685 + 7 * 55.84514413384 + 2 * 24.30505162813 + 3 * 10.811 + 31 * 15.999404924748 + 6 * 28.0855 + 3 * 1.0079407540258,
            'Ca[(Fe3+)2Fe2+][(Fe3+)4Mg2](BO3)3Si6O18(OH)3O' => 40.078022556685 + 7 * 55.84514413384 + 2 * 24.30505162813 + 3 * 10.811 + 31 * 15.999404924748 + 6 * 28.0855 + 3 * 1.0079407540258,
            'Na2Fe2+(CaNa2)(Fe2+)13Al[(PO4)11(PO3OH)(OH)2]' => 2 * 22.98976928 + 55.84514413384 + 40.078022556685 + 2 * 22.98976928 + 13 * 55.84514413384 + 26.9815386 + 11 * (30.973761998 + 4 * 15.999404924748) + 30.973761998 + 3 * 15.999404924748 + 15.999404924748 + 1.0079407540258 + 2 * (15.999404924748 + 1.0079407540258),
            'Mg6Cr2CO3(OH)16 · 4H2O' => 6 * 24.30505162813 + 2 * 51.9961 + 12.010735896734 + 3 * 15.999404924748 + 16 * (15.999404924748 + 1.0079407540258) + 4 * (2 * 1.0079407540258 + 15.999404924748),
            '(NH4)2(UO2)2(PO4)2 · 6H2O' => 2 * (14.006703211018 + 4 * 1.0079407540258) + 2 * (238.02914609015 + 2 * 15.999404924748) + 2 * (30.973761998 + 4 * 15.999404924748) + 6 * (2 * 1.0079407540258 + 15.999404924748),
            'Pb4,5Sb4,5S11' => 4.5 * 207.216908331 + 4.5 * 121.75978116 + 11 * 32.064787405608,
            'Na(Li1,5Al1,5)Al6(BO3)3Si6O18(OH)3F' => 22.98976928 + 1.5 * 6.9400366060273 + 1.5 * 26.9815386 + 6 * 26.9815386 + 3 * (10.811 + 3 * 15.999404924748) + 6 * 28.0855 + 18 * 15.999404924748 + 3 * (15.999404924748 + 1.0079407540258) + 18.9984032,
            'Na12(K,Sr,Ce)3Ca6Mn3Zr3NbSi25O73(O,H2O,OH)5(OH,F,Cl)2' => 4116.3502809157,
            'Mn2+(Ti,Nb)5O12 · 9H2O' => 54.938044 + 5 * (47.8667450392 + 92.90637) + 12 * 15.999404924748 + 9 * (2 * 1.0079407540258 + 15.999404924748),
            'Ca(Fe3+)2(PO4)2(OH,F)2' => 40.078022556685 + 2 * 55.84514413384 + 2 * (30.973761998 + 4 * 15.999404924748) + 2 * (15.999404924748 + 1.0079407540258 + 18.9984032),
            '(Ba,Na)2(Na,Ti,Mn)4(Ti,Nb)2Si4O14(OH,O,F)5 · 3H2O' => 2 * (137.3266716329 + 22.98976928) + 4 * (22.98976928 + 47.8667450392 + 54.938044) + 2 * (47.8667450392 + 92.90637) + 4 * 28.0855 + 14 * 15.999404924748 + 5 * (15.999404924748 + 1.0079407540258 + 15.999404924748 + 18.9984032) + 3 * (2 * 1.0079407540258 + 15.999404924748),
            'C15-H13-Cl-N-Na-O3.2H2O' => 15 * 12.010735896734 + 13 * 1.0079407540258 + 35.453 + 14.006703211018 + 22.98976928 + 3 * 15.999404924748 + 2 * (2 * 1.0079407540258 + 15.999404924748),
            'C15-H13-Cl-N-Na-O3.2H2-O' => 15 * 12.010735896734 + 13 * 1.0079407540258 + 35.453 + 14.006703211018 + 22.98976928 + 3 * 15.999404924748 + 2 * (2 * 1.0079407540258 + 15.999404924748),
        ];
    }
    
    /**
     * @dataProvider getDataForTestItCanDetectInvalidChemicalFormula
     * @expectedException \FiiSoft\MolecularWeightCalculator\Exception\InvalidChemicalFormulaException
     */
    public function test_it_can_detect_invalid_chemical_formula($chemicalFormula)
    {
        $this->calc->computeWeight($chemicalFormula);
    }
    
    public function getDataForTestItCanDetectInvalidChemicalFormula()
    {
        return [
            ['MnOp3'],
            ['COZ'],
            ['Ch3(CO2)('],
            ['()Na'],
            ['H2O5(S)Na)'],
        ];
    }
    
    public function test_compute_weight_from_data_structure()
    {
        $structure = [
            'type' => 'group',
            'quantity' => 1,
            'items' => [
                [
                    'type' => 'leaf',
                    'symbol' => 'K',
                    'mass' => 39.0983,
                ], [
                    'type' => 'group',
                    'quantity' => 2,
                    'items' => [
                        [
                            'type' => 'leaf',
                            'symbol' => 'C2',
                            'mass' => 2 * 12.010735896734,
                        ], [
                            'type' => 'group',
                            'quantity' => 1,
                            'items' => [
                                [
                                    'type' => 'leaf',
                                    'symbol' => 'O5',
                                    'mass' => 5 * 15.999404924748,
                                ]
                            ]
                        ], [
                            'type' => 'group',
                            'quantity' => 3,
                            'items' => [
                                [
                                    'type' => 'leaf',
                                    'symbol' => 'P2',
                                    'mass' => 2 * 30.973761998,
                                ]
                            ]
                        ]
                    ],
                ], [
                    'type' => 'leaf',
                    'symnol' => 'S',
                    'mass' => 32.064787405608,
                ]
            ],
        ];
        
        $compute = null;
        $compute = function (array $node) use (&$compute) {
            if ($node['type'] === 'group') {
                return $node['quantity'] * array_sum(array_map($compute, $node['items']));
            }
            return $node['mass'];
        };
    
        $mass = $compute($structure);
        
        self::assertTrue(abs(650.885244 - $mass) < 0.001);
    }
}


src/MolecularWeightCalculator.php
<?php

namespace FiiSoft\MolecularWeightCalculator;

use FiiSoft\MolecularWeightCalculator\Exception\CorruptedDataException;
use FiiSoft\MolecularWeightCalculator\Exception\InvalidChemicalFormulaException;
use FiiSoft\MolecularWeightCalculator\Exception\UnableToComputeMolecularWeightException;
use InvalidArgumentException;

final class MolecularWeightCalculator
{
    const SIMPLE_FORMULA = '/^[\d\p{L},]+$/u';
    const FORMULA_WITH_GROUPS = '/^[,\d\p{L}\(\)\[\]]+$/u';
    const STARTS_WITH_DIGITS = '/^\s*(?P<number>\d+)/u';
    const FORMULA_WITH_IONS = '/\d+(\+|\-)/u';
    const COMPOUND_FORMULA = '/(?<!\.)(·|\.)(?!\.)/u';
    
    /** @var string */
    private $symbolAndQuantityPattern;
    
    /** @var array */
    private $data = [];
    
    /** @var array */
    private $masses = [];
    
    /**
     * @throws CorruptedDataException when read data failed
     */
    public function __construct()
    {
        $this->init();
    }
    
    /**
     * @throws CorruptedDataException
     * @return void
     */
    private function init()
    {
        $this->loadDataSetOne();
        $this->loadDataSetTwo();
    }
    
    /**
     * @throws CorruptedDataException
     * @return void
     */
    private function loadDataSetOne()
    {
        $dataFile = implode(DIRECTORY_SEPARATOR, [__DIR__, '..', 'data', 'atoms_weights_data.php']);
        
        $data = include $dataFile;
        $numOfItems = 4 * 112; //112 elements in data
        
        if (!is_array($data) || count($data) !== $numOfItems) {
            $this->throwExceptionCorruptedData($dataFile);
        }
    
        $prevAtomicNumber = 1;
        
        for ($i = 0; $i < $numOfItems; $i += 4) {
            if (!isset($data[$i], $data[$i+1], $data[$i+2], $data[$i+3])) {
                $this->throwExceptionCorruptedData($dataFile);
            }
            
            $atomicNumber = (int) $data[$i];
            $name = trim($data[$i+1]);
            $symbol = trim($data[$i+2]);
            $atomicMass = (float) $data[$i+3];
    
            if ($atomicNumber !== $prevAtomicNumber++ || $name === '' || $symbol === '' || $atomicMass < 1.0
                || isset($this->data[$symbol])
            ) {
                $this->throwExceptionCorruptedData($dataFile);
            }
            
            $this->data[$symbol] = [
                'number' => $atomicNumber,
                'name' => $name,
                'symbol' => $symbol,
                'mass' => $atomicMass,
            ];
            
            $this->masses[$symbol] = $atomicMass;
        }
    
        $symbols = array_keys($this->masses);
        rsort($symbols);
    
        $this->symbolAndQuantityPattern = str_replace(
            '{symbols}', implode('|', $symbols), '/(?P<symbol>{symbols})(?P<quantity>[\d,]+)?/u'
        );
    }
    
    /**
     * @throws CorruptedDataException
     * @return void
     */
    private function loadDataSetTwo()
    {
        $dataFile = implode(DIRECTORY_SEPARATOR, [__DIR__, '..', 'data', 'isotopesmatter.txt']);
    
        foreach (file($dataFile) as $line) {
            $line = trim($line);
            if ($line === '') {
                break;
            }
            
            list($symbol, $mass) = explode(': ', $line);
            $mass = (float) $mass;
            
            $this->data[$symbol]['mass'] = $mass;
            $this->masses[$symbol] = $mass;
        }
    }
    
    /**
     * @param string $chemicalFormula
     * @throws UnableToComputeMolecularWeightException
     * @throws InvalidChemicalFormulaException
     * @throws InvalidArgumentException
     * @return float
     */
    public function computeWeight($chemicalFormula)
    {
        $chemicalFormula = is_string($chemicalFormula) ? trim($chemicalFormula) : '';
        if ($chemicalFormula === '') {
            throw new InvalidArgumentException('Invalid param chemicalFormula');
        }
        
        $weight = $this->computeFromSimpleFormula($chemicalFormula)
            ?: $this->computeFromFormulaWithGroups($chemicalFormula)
            ?: $this->computeFromFormulaWithDashes($chemicalFormula)
            ?: $this->computeFromFormulaWithIons($chemicalFormula)
            ?: $this->computeFromCompoundFormula($chemicalFormula);
        
        if ($weight !== false) {
            return $weight;
        }
        
        $this->throwExceptionUnableToComputeWeight($chemicalFormula);
    }
    
    /**
     * @param string $chemicalFormula
     * @throws UnableToComputeMolecularWeightException
     * @throws InvalidChemicalFormulaException
     * @throws InvalidArgumentException
     * @return float|false
     */
    private function computeFromCompoundFormula($chemicalFormula)
    {
        if (preg_match(self::COMPOUND_FORMULA, $chemicalFormula)) {
            $formulas = preg_split(self::COMPOUND_FORMULA, $chemicalFormula, -1, PREG_SPLIT_NO_EMPTY);
            if (is_array($formulas)) {
                return array_sum(array_map(function ($formula) use ($chemicalFormula) {
                    if (preg_match(self::STARTS_WITH_DIGITS, $formula, $matches) && isset($matches['number'])) {
                        $number = $this->asFloat($matches['number']);
                        $formula = preg_replace(self::STARTS_WITH_DIGITS, '', $formula);
                    } else {
                        $number = 1.0;
                    }
                    
                    $weight = $this->computeWeight($formula);
                    if ($weight !== false) {
                        return $number === 1.0 ? $weight : $weight * $number;
                    }
                    
                    $this->throwExceptionUnableToComputeWeight($chemicalFormula);
                }, $formulas));
            }
        }
        
        return false;
    }
    
    /**
     * @param string $chemicalFormula
     * @throws UnableToComputeMolecularWeightException
     * @throws InvalidChemicalFormulaException
     * @throws InvalidArgumentException
     * @return float|false
     */
    private function computeFromFormulaWithIons($chemicalFormula)
    {
        return preg_match(self::FORMULA_WITH_IONS, $chemicalFormula)
            ? $this->computeWeight(preg_replace(self::FORMULA_WITH_IONS, '', $chemicalFormula))
            : false;
    }
    
    /**
     * @param string $chemicalFormula
     * @throws UnableToComputeMolecularWeightException
     * @throws InvalidChemicalFormulaException
     * @throws InvalidArgumentException
     * @return float|false
     */
    private function computeFromFormulaWithDashes($chemicalFormula)
    {
        return false !== strpos($chemicalFormula, '-')
            ? $this->computeWeight(str_replace('-', '', $chemicalFormula))
            : false;
    }
    
    /**
     * @param string $chemicalFormula
     * @throws InvalidChemicalFormulaException
     * @return float|false
     */
    private function computeFromFormulaWithGroups($chemicalFormula)
    {
        if (!preg_match(self::FORMULA_WITH_GROUPS, $chemicalFormula)) {
            return false;
        }
    
        $formula = str_replace(['[', ']'], ['(', ')'], $chemicalFormula);
        
        $pieces = preg_split('/(\(|\))/u', $formula, -1, PREG_SPLIT_DELIM_CAPTURE | PREG_SPLIT_NO_EMPTY);
        if (!is_array($pieces)) {
            return false;
        }
    
        $numOfGroups = substr_count($formula, '(');
        if ($numOfGroups !== substr_count($formula, ')')) {
            $this->throwExceptionInvalidFormula($chemicalFormula);
        }
        
        $groupPrototype = [
            'type' => 'group',
            'quantity' => 1.0,
            'items' => [],
        ];
        
        $level = 0;
        $levels = [];
        $structure = $groupPrototype;
        $levels[$level] = &$structure;
        
        for ($i = 0, $max = count($pieces); $i < $max; ++$i) {
            $item = $pieces[$i];
            if ($item === '(') {
                unset($group);
                $group = $groupPrototype;
                $levels[$level]['items'][] = &$group;
                $levels[++$level] = &$group;
            } elseif ($item === ')') {
                if ($level > 0) {
                    $number = 1;
                    if ($i + 1 < $max) {
                        $nextItem = $pieces[$i + 1];
                        if (ctype_digit($nextItem)) {
                            $number = $this->asFloat($nextItem);
                            ++$i;
                        } else {
                            $matches = [];
                            if (preg_match(self::STARTS_WITH_DIGITS, $nextItem, $matches) && isset($matches['number'])) {
                                $number = $this->asFloat($matches['number']);
                                $pieces[$i + 1] = preg_replace(self::STARTS_WITH_DIGITS, '', $nextItem);
                            }
                        }
                    }
                    $levels[$level--]['quantity'] = $number;
                } else {
                    $this->throwExceptionInvalidFormula($chemicalFormula);
                }
            } else {
                $mass = $this->computeFromSimpleFormula($item);
                if ($mass !== false) {
                    $levels[$level]['items'][] = [
                        'type' => 'leaf',
                        'item' => $item,
                        'mass' => $mass,
                    ];
                } else {
                    $this->throwExceptionInvalidFormula($chemicalFormula);
                }
            }
        }
    
        $compute = null;
        $compute = function (array $node) use (&$compute, $chemicalFormula) {
            if ($node['type'] === 'group') {
                if (empty($node['items'])) {
                    $this->throwExceptionInvalidFormula($chemicalFormula);
                }
                return $node['quantity'] * array_sum(array_map($compute, $node['items']));
            }
            return $node['mass'];
        };

        return $compute($structure);
    }
    
    /**
     * @param string $chemicalFormula
     * @throws InvalidChemicalFormulaException
     * @return float|false
     */
    private function computeFromSimpleFormula($chemicalFormula)
    {
        if (!preg_match(self::SIMPLE_FORMULA, $chemicalFormula)) {
            return false;
        }
    
        $formula = preg_replace('/(?<=\p{L}),(?=\p{L})/u', '', $chemicalFormula);
        if (!preg_match_all($this->symbolAndQuantityPattern, $formula, $matches, PREG_SPLIT_DELIM_CAPTURE)) {
            return false;
        }
        
        $totalMass = 0.0;
    
        foreach ($matches as $match) {
            if (0 === strpos($formula, $match['symbol'])) {
                $formula = substr($formula, strlen($match['symbol']));
            } else {
                $this->throwExceptionInvalidFormula($chemicalFormula);
            }
    
            if (isset($match['quantity'])) {
                if (0 === strpos($formula, $match['quantity'])) {
                    $formula = substr($formula, strlen($match['quantity']));
                } else {
                    $this->throwExceptionInvalidFormula($chemicalFormula);
                }
                
                $quantity = $this->asFloat($match['quantity']);
            } else {
                $quantity = 1.0;
            }
            
            $totalMass += $quantity * $this->masses[$match['symbol']];
        }
    
        if ($formula !== false && $formula !== '') {
            $this->throwExceptionInvalidFormula($chemicalFormula);
        }
        
        return $totalMass;
    }
    
    /**
     * @param string $value
     * @return float
     */
    private function asFloat($value)
    {
        return (float) str_replace(',', '.', $value);
    }
    
    /**
     * @return array
     */
    public function getData()
    {
        return $this->data;
    }
    
    /**
     * @param string $chemicalFormula
     * @throws InvalidChemicalFormulaException
     * @return void
     */
    private function throwExceptionInvalidFormula($chemicalFormula)
    {
        throw InvalidChemicalFormulaException::invalidFormula($chemicalFormula);
    }
    
    /**
     * @param string $chemicalFormula
     * @throws UnableToComputeMolecularWeightException
     * @return void
     */
    private function throwExceptionUnableToComputeWeight($chemicalFormula)
    {
        throw UnableToComputeMolecularWeightException::fromFormula($chemicalFormula);
    }
    
    /**
     * @param string $dataFile
     * @throws CorruptedDataException
     * @return void
     */
    private function throwExceptionCorruptedData($dataFile)
    {
        throw CorruptedDataException::inFile($dataFile);
    }
}

src/atoms_weights_data.php
<?php
return [
    '1',
    'Hydrogen',
    'H',
    '1.00794',
    '2',
    'Helium',
    'He',
    '4.002602',
    '3',
    'Lithium',
    'Li',
    '6.941',
    '4',
    'Beryllium',
    'Be',
    '9.012182',
    '5',
    'Boron',
    'B',
    '10.811',
    '6',
    'Carbon',
    'C',
    '12.0107',
    '7',
    'Nitrogen',
    'N',
    '14.0067',
    '8',
    'Oxygen',
    'O',
    '15.9994',
    '9',
    'Fluorine',
    'F',
    '18.9984032',
    '10',
    'Neon',
    'Ne',
    '20.1797',
    '11',
    'Sodium',
    'Na',
    '22.98976928',
    '12',
    'Magnesium',
    'Mg',
    '24.3050',
    '13',
    'Aluminum',
    'Al',
    '26.9815386',
    '14',
    'Silicon',
    'Si',
    '28.0855',
    '15',
    'Phosphorus',
    'P',
    '30.973762',
    '16',
    'Sulfur',
    'S',
    '32.065',
    '17',
    'Chlorine',
    'Cl',
    '35.453',
    '18',
    'Argon',
    'Ar',
    '39.948',
    '19',
    'Potassium',
    'K',
    '39.0983',
    '20',
    'Calcium',
    'Ca',
    '40.078',
    '21',
    'Scandium',
    'Sc',
    '44.955912',
    '22',
    'Titanium',
    'Ti',
    '47.867',
    '23',
    'Vanadium',
    'V',
    '50.9415',
    '24',
    'Chromium',
    'Cr',
    '51.9961',
    '25',
    'Manganese',
    'Mn',
    '54.938045',
    '26',
    'Iron',
    'Fe',
    '55.845',
    '27',
    'Cobalt',
    'Co',
    '58.933195',
    '28',
    'Nickel',
    'Ni',
    '58.6934',
    '29',
    'Copper',
    'Cu',
    '63.546',
    '30',
    'Zinc',
    'Zn',
    '65.409',
    '31',
    'Gallium',
    'Ga',
    '69.723',
    '32',
    'Germanium',
    'Ge',
    '72.64',
    '33',
    'Arsenic',
    'As',
    '74.92160',
    '34',
    'Selenium',
    'Se',
    '78.96',
    '35',
    'Bromine',
    'Br',
    '79.904',
    '36',
    'Krypton',
    'Kr',
    '83.798',
    '37',
    'Rubidium',
    'Rb',
    '85.4678',
    '38',
    'Strontium',
    'Sr',
    '87.62',
    '39',
    'Yttrium',
    'Y',
    '88.90585',
    '40',
    'Zirconium',
    'Zr',
    '91.224',
    '41',
    'Niobium',
    'Nb',
    '92.90638',
    '42',
    'Molybdenum',
    'Mo',
    '95.94',
    '43',
    'Technetium',
    'Tc',
    '98.9063',
    '44',
    'Ruthenium',
    'Ru',
    '101.07',
    '45',
    'Rhodium',
    'Rh',
    '102.90550',
    '46',
    'Palladium',
    'Pd',
    '106.42',
    '47',
    'Silver',
    'Ag',
    '107.8682',
    '48',
    'Cadmium',
    'Cd',
    '112.411',
    '49',
    'Indium',
    'In',
    '114.818',
    '50',
    'Tin',
    'Sn',
    '118.710',
    '51',
    'Antimony',
    'Sb',
    '121.760',
    '52',
    'Tellurium',
    'Te',
    '127.60',
    '53',
    'Iodine',
    'I',
    '126.90447',
    '54',
    'Xenon',
    'Xe',
    '131.293',
    '55',
    'Caesium',
    'Cs',
    '132.9054519',
    '56',
    'Barium',
    'Ba',
    '137.327',
    '57',
    'Lanthanum',
    'La',
    '138.90547',
    '58',
    'Cerium',
    'Ce',
    '140.116',
    '59',
    'Praseodymium',
    'Pr',
    '140.90765',
    '60',
    'Neodymium',
    'Nd',
    '144.242',
    '61',
    'Promethium',
    'Pm',
    '146.9151',
    '62',
    'Samarium',
    'Sm',
    '150.36',
    '63',
    'Europium',
    'Eu',
    '151.964',
    '64',
    'Gadolinium',
    'Gd',
    '157.25',
    '65',
    'Terbium',
    'Tb',
    '158.92535',
    '66',
    'Dysprosium',
    'Dy',
    '162.500',
    '67',
    'Holmium',
    'Ho',
    '164.93032',
    '68',
    'Erbium',
    'Er',
    '167.259',
    '69',
    'Thulium',
    'Tm',
    '168.93421',
    '70',
    'Ytterbium',
    'Yb',
    '173.04',
    '71',
    'Lutetium',
    'Lu',
    '174.967',
    '72',
    'Hafnium',
    'Hf',
    '178.49',
    '73',
    'Tantalum',
    'Ta',
    '180.9479',
    '74',
    'Tungsten',
    'W',
    '183.84',
    '75',
    'Rhenium',
    'Re',
    '186.207',
    '76',
    'Osmium',
    'Os',
    '190.23',
    '77',
    'Iridium',
    'Ir',
    '192.217',
    '78',
    'Platinum',
    'Pt',
    '195.084',
    '79',
    'Gold',
    'Au',
    '196.966569',
    '80',
    'Mercury',
    'Hg',
    '200.59',
    '81',
    'Thallium',
    'Tl',
    '204.3833',
    '82',
    'Lead',
    'Pb',
    '207.2',
    '83',
    'Bismuth',
    'Bi',
    '208.98040',
    '84',
    'Polonium',
    'Po',
    '208.9824',
    '85',
    'Astatine',
    'At',
    '209.9871',
    '86',
    'Radon',
    'Rn',
    '222.0176',
    '87',
    'Francium',
    'Fr',
    '223.0197',
    '88',
    'Radium',
    'Ra',
    '226.0254',
    '89',
    'Actinium',
    'Ac',
    '227.0278',
    '90',
    'Thorium',
    'Th',
    '232.03806',
    '91',
    'Protactinium',
    'Pa',
    '231.03588',
    '92',
    'Uranium',
    'U',
    '238.02891',
    '93',
    'Neptunium',
    'Np',
    '237.0482',
    '94',
    'Plutonium',
    'Pu',
    '244.0642',
    '95',
    'Americium',
    'Am',
    '243.0614',
    '96',
    'Curium',
    'Cm',
    '247.0703',
    '97',
    'Berkelium',
    'Bk',
    '247.0703',
    '98',
    'Californium',
    'Cf',
    '251.0796',
    '99',
    'Einsteinium',
    'Es',
    '252.0829',
    '100',
    'Fermium',
    'Fm',
    '257.0951',
    '101',
    'Mendelevium',
    'Md',
    '258.0986',
    '102',
    'Nobelium',
    'No',
    '259.1009',
    '103',
    'Lawrencium',
    'Lr',
    '260.1053',
    '104',
    'Rutherfordium',
    'Rf',
    '261.1087',
    '105',
    'Dubnium',
    'Db',
    '262.1138',
    '106',
    'Seaborgium',
    'Sg',
    '263.1182',
    '107',
    'Bohrium',
    'Bh',
    '262.1229',
    '108',
    'Hassium',
    'Hs',
    '265',
    '109',
    'Meitnerium',
    'Mt',
    '266',
    '110',
    'Darmstadtium',
    'Ds',
    '269',
    '111',
    'Roentgenium',
    'Rg',
    '272',
    '112',
    'Copernicium',
    'Cn',
    '285',
];